### ISSUE TITLE
How do I "run" my python dictionary to get it to output a json?

### ISSUE THREAD

**Issue by winstond** [user202729/plover-python-dictionary-lib#4]
I tried:
python dictionary.py

But I get an error: ModuleNotFoundError: No module named 'plover'.   I tried a few things, but I'm not super familiar with python so not sure what I'm missing.

Thank you.

**Comment by user202729** [comment]
Not sure what I can support about this. I think you have a few options

* hard code the system instead of importing from Plover
* run it with a Python executable with access to Plover (depends on the operating system the instruction might be different... are you on Windows/Linux/Mac?)

**Comment by winstond** [comment]
Yes, I do apologize.  More of a Python question really.  I'm on Win 11.  Plover is in the usual place, C:\\Program Files\Open Steno Project\Plover.

Perhaps once I can confirm the way to do this it could be added to the readme page?

I have a dictionary for Hindi I'd like to share, but right now I'd have to provide command line instructions for people to use it.  Hoping to just have a JSON.

**Comment by winstond** [comment]
Update: I managed to get it to work.  Here are the steps in case anyone else stumbles on this:

Windows 11.  Python install on system, but Plover uses it's own...

1. Open a powershell (might need to open as admin depending on where your files are)
2. Navigate to your plover directory (ex: c:\program files\open steno project\plover)
3. Install plover-python-dictionary-lib in this instance of python with:
.\data\python.exe -m pip install plover_python_dictionary_lib
4. Run your python dictionary (which should include the line to print the dictionary to the terminal as per readme)
 .\data\python.exe .\hindirun.py
Your dictionary should dump out to the terminal
5. Highlight and copy your dictionary from the command prompt and make your json file.

(I tried piping to a file, but my JSON output had unicode characters so I got an encoding error.  However copying the output in the terminal and pasting it still worked even though the terminal wouldn't display the the unicode characters properly.)

**Comment by user202729** [comment]
Some day it would be easier.

 https://github.com/openstenoproject/plover_python_dictionary/issues/9

**Comment by winstond** [comment]
Having an issue with the above steps.  Did something change in this library?  I tried the steps above and I'm now getting a "Duplicate Keys" error.

More detail:
I have a .py file.  When loaded into Plover as a python dictionary it works fine.  However when I attempt to print out the output to json following the steps I previously documented in this issue, I now am getting an error "duplicate keys".

What I'm confused about is if there was a problem with the .py dictionary file I would expect to get an error when the dictionary is loaded in plover.  I'm not getting any error there.  I'm only getting this error when I attempt to output to a json file in a command window.

Here's the code error from my command prompt:
Traceback (most recent call last):
  File ".\hindirun.py", line 1123, in <module>
    dictionary.print_items()
  File "C:\Program Files\Open Steno Project\Plover_continuous\data\lib\site-packages\plover_python_dictionary_lib\__init__.py", line 112, in print_items
    json.dump(self.items_str_dict(), sys.stdout, ensure_ascii=False, indent=0)
  File "C:\Program Files\Open Steno Project\Plover_continuous\data\lib\site-packages\plover_python_dictionary_lib\__init__.py", line 104, in items_str_dict
    assert len(result_items)==len(result), "Duplicate keys"
AssertionError: Duplicate keys

**Comment by user202729** [comment]
The problem is it's nontrivial to **detect** duplicate keys quickly. In order to detect it the program can list all the key-value pairs, but if it does every time Plover starts it would be slow. On the other hand when printing out the JSON dictionary, it needs to list all the key/values anyway, so it might as well do.

**Comment by winstond** [comment]
I guess I don't understand why that control about duplicates is needed given that Plover doesn't fail with the same .py file.

That said, if the error could tell me what the duplicate was I could attempt to modify the .py file to ensure my python logic doesn't produce any duplicate stroke definitions.

As it is, I don't get the list, just the error.  And no feedback on what I would need to modify to avoid the error and get a print out.

**Comment by user202729** [comment]
I modified it to just print a warning of some example duplicating entries instead of an error.

**Comment by winstond** [comment]
Oh awesome!  Thank you.

**Comment by winstond** [comment]
Just coming back on this... I didn't really have time to test it, but was just able to do so.

1. I'm now able to follow the instructions I posted in a previous comment successfully (I do now get the warning with a list of duplicate strokes, but then I also get the JSON output!)
Instructions: https://github.com/user202729/plover-python-dictionary-lib/issues/4#issuecomment-2430179682
2. Base on the duplicate stroke list it took less than 5 minutes to delete the offending line in my .py file to elimate the duplicates.

Yay!

Now if only we could just save the JSON file directly from the Plover window ;)

**Comment by user202729** [comment]
In principle, it isn't too difficult. Proof of concept:

1. Change your Python dictionary file as follows

	```diff
	 LONGEST_KEY = dictionary.longest_key
	+items = lambda: [(tuple(str(s) for s in k), v) for k, v in dictionary.items()]
	```

2. Install my fork of plover-python-dictionary https://github.com/user202729/plover_python_dictionary
3. Right click and "save dictionaries as" etc.

In practice, step 2 is the difficult part… I made an (unrelated) pull request on plover_python_dictionary https://github.com/openstenoproject/plover_python_dictionary/pull/11 a few months ago, and nobody notices.

(caveat is the warnings won't be too visible if you do it like this. There isn't much I can do about this, I suppose.)

**Comment by winstond** [comment]
I guess now that I know how to it from command line, that's probably better.  The warning was helpful.  Thanks!
